chronos/streams/tlsstream.nim

```diff
@@ -0,0 +1,334 @@
+
+  finally:
+    if wstream.state == AsyncStreamState.Stopped:
+      while len(wstream.queue) > 0:
+        let item = wstream.queue.popFirstNoWait()
+        if not(item.future.finished()):
+          item.future.complete()
+    elif wstream.state == AsyncStreamState.Error:
+      while len(wstream.queue) > 0:
+        let item = wstream.queue.popFirstNoWait()
+        if not(item.future.finished()):
+          item.future.fail(error)
+    wstream.stream = nil
+
+proc tlsReadLoop(stream: AsyncStreamReader) {.async.} =
+  var rstream = cast[TlsStreamReader](stream)
+  var engine: ptr SslEngineContext
+
+  if rstream.kind == TlsStreamKind.Server:
+    engine = addr rstream.scontext.eng
+  else:
+    engine = addr rstream.ccontext.eng
+
+  rstream.state = AsyncStreamState.Running
+
+  try:
+    var length: uint
+    while true:
+      var state = engine.sslEngineCurrentState()
+
+      if (state and SSL_CLOSED) == SSL_CLOSED:
+        let err = engine.sslEngineLastError()
+        if err != 0:
+          rstream.error = newTlsStreamProtocolError(err)
+          rstream.state = AsyncStreamState.Error
+          break
+        else:
+          rstream.state = AsyncStreamState.Stopped
+          break
+
+      if (state and (SSL_SENDREC or SSL_SENDAPP)) != 0:
+        if not(rstream.switchToWriter.isSet()):
+          rstream.switchToWriter.fire()
+
+      if (state and (SSL_RECVREC or SSL_RECVAPP)) == 0:
+        await rstream.switchToReader.wait()
+        rstream.switchToReader.clear()
+        # We need to refresh `state` because we just returned from writerLoop.
+        continue
+
+      if (state and SSL_RECVREC) == SSL_RECVREC:
+        # TLS records required for further processing
+        length = 0'u
+        var buf = sslEngineRecvrecBuf(engine, length)
+        var resFut = awaitne rstream.rsource.readOnce(buf, int(length))
+        if resFut.failed():
+          rstream.error = resFut.error
+          rstream.state = AsyncStreamState.Error
+          break
+        let res = resFut.read()
+        if res > 0:
+          sslEngineRecvrecAck(engine, uint(res))
+          continue
+        else:
+          rstream.state = AsyncStreamState.Finished
+          break
+
+      if (state and SSL_RECVAPP) == SSL_RECVAPP:
+        # Application data can be recovered.
+        length = 0'u
+        var buf = sslEngineRecvappBuf(engine, length)
+        await upload(addr rstream.buffer, buf, int(length))
+        sslEngineRecvappAck(engine, length)
+        continue
+
+  except CancelledError:
+    rstream.state = AsyncStreamState.Stopped
+
+  finally:
+    # Perform TLS cleanup procedure
+    sslEngineClose(engine)
+    # Becase tlsWriteLoop() is ephemeral, but we still need to keep stream state
+    # consistent.
+    rstream.buffer.forget()
+    rstream.stream = nil
+
+proc newTlsClientAsyncStream*(rsource: AsyncStreamReader,
+                              wsource: AsyncStreamWriter,
+                              serverName: string = "",
```

Comment:
Since it's generally a bad idea to provide empty server name, this argument should be mandatory.

Reply:
TLS streams are used not only in TCP/IP network, because we made it as a stream it can be used for every possible types of transports (e.g. pipes, files). In such use-case server name is not available by default. One more reason to leave it is to allow people, who knows little about SSL/TLS protocols to create encrypted communication channel without any problems.

Reply:
The *default* we want here is that users provide the server name, with an option to not provide it. That is exactly the opposite of what default argument values are. "People who know little" is an invalid argument. This is security related API, and thus security/correctness should always be favored over simplicity.